File: web/api/routers/tracks.py
"""Tracks router — list/search all tracks ever played on the show."""

from fastapi import APIRouter, Request, HTTPException, Query
from ..models.schemas import TrackCount

import sqlite3


router = APIRouter(tags=["tracks"])


def _db(request: Request) -> sqlite3.Connection:
    db = sqlite3.connect(request.app.state.db_path)
    db.row_factory = sqlite3.Row
    return db


@router.get("")
def list_tracks(
    request: Request,
    page: int = Query(1, ge=1),
    per_page: int = Query(50, ge=1, le=200),
    sort: str = Query("-plays"),
    search: str = Query(""),
):
    """List all tracks with play statistics."""
    db = _db(request)
    try:
        offset = (page - 1) * per_page

        where_clauses = ["1=1"]
        params = []

        if search:
            where_clauses.append("(t.title LIKE ? OR t.artist LIKE ?)")
            params.extend([f"%{search}%", f"%{search}%"])

        where_sql = "WHERE " + " AND ".join(where_clauses)

        order_dir = "DESC" if sort.startswith("-") else "ASC"
        order_col = sort.lstrip("-")
        allowed_cols = {"plays", "title", "artist", "last_played", "album"}
        if order_col not in allowed_cols:
            order_col = "plays"

        # We group by title, artist, and album to ensure different versions are counted separately.
        # This accurately reflects the user's requirement that plays are per album.

        total = db.execute(
            f"""SELECT COUNT(*) as c FROM (
                SELECT t.title, t.artist_id, t.album_id FROM tracks t {where_sql}
                GROUP BY t.title, t.artist_id, t.album_id
            )""",
            params,
        ).fetchone()["c"]

        rows = db.execute(
            f"""SELECT t.title, art.name as artist, alb.name as album,
                       COUNT(*) as plays,
                       COUNT(DISTINCT t.episode_id) as episodes,
                       MAX(e.date) as last_played
                FROM tracks t
                JOIN episodes e ON e.id = t.episode_id
                JOIN artists art ON t.artist_id = art.id
                LEFT JOIN albums alb ON t.album_id = alb.id
                {where_sql}
                GROUP BY t.title, t.artist_id, t.album_id
                ORDER BY {order_col} {order_dir}
                LIMIT ? OFFSET ?""",
            (*params, per_page, offset),
        ).fetchall()

        return {
            "items": [
                TrackCount(
                    title=r["title"],
                    plays=r["plays"],
                    episodes=r["episodes"],
                    last_played=r["last_played"],
                    artist=r["artist"],
                    album=r["album"],
                )
                for r in rows
            ],
            "total": total,
            "page": page,
            "per_page": per_page,
        }
    finally:
        db.close()

File: worker/src/routers/tracks.ts
/**
 * Tracks router — list/search all tracks ever played on the show.
 */

import { Hono } from 'hono'
import type { Env } from '../db'
import { DB } from '../db'

export const tracksRouter = new Hono<{ Bindings: Env }>()

tracksRouter.get('/', async (c) => {
  const db = DB(c.env)
  const page = Math.max(1, Number(c.req.query('page')) || 1)
  const perPage = Math.max(1, Math.min(200, Number(c.req.query('per_page')) || 50))
  const sort = c.req.query('sort') || '-plays'
  const search = c.req.query('search') || ''
  const offset = (page - 1) * perPage

  const whereClauses: string[] = ['1=1']
  const params: unknown[] = []

  if (search) {
    whereClauses.push('(t.title LIKE ? OR t.artist LIKE ?)')
    params.push(`%${search}%`, `%${search}%`)
  }

  const whereSQL = 'WHERE ' + whereClauses.join(' AND ')
  const orderDir = sort.startsWith('-') ? 'DESC' : 'ASC'
  const orderCol = sort.replace(/^-/, '')

  const total = (await db.one<{ c: number }>(
    `SELECT COUNT(*) as c FROM (
      SELECT t.title, t.artist_id, t.album_id FROM tracks t ${whereSQL}
      GROUP BY t.title, t.artist_id, t.album_id
    )`,
    ...params,
  ))?.c ?? 0

  const { results: rows } = await db.all<{
    title: string; artist: string; album: string | null
    plays: number; episodes: number; last_played: string | null
  }>(
    `SELECT t.title, art.name as artist, alb.name as album,
            COUNT(*) as plays,
            COUNT(DISTINCT t.episode_id) as episodes,
            MAX(e.date) as last_played
     FROM tracks t
     JOIN episodes e ON e.id = t.episode_id
     JOIN artists art ON t.artist_id = art.id
     LEFT JOIN albums alb ON t.album_id = alb.id
     ${whereSQL}
     GROUP BY t.title, t.artist_id, t.album_id
     ORDER BY ${orderCol} ${orderDir}
     LIMIT ? OFFSET ?`,
    ...params, perPage, offset,
  )

  return c.json({
    items: (rows ?? []).map((r) => ({
      title: r.title,
      plays: r.plays,
      episodes: r.episodes,
      last_played: r.last_played ?? null,
      artist: r.artist,
      album: r.album ?? null,
    })),
    total,
    page,
    per_page: perPage,
  })
})
